tools/nemo_forced_aligner/utils/make_output_files.py

```diff
@@ -0,0 +1,210 @@
+        {'text': 'hey', 's_start': 19, 's_end': 23},
+    ]
+
+    and the alignment could be
+    alignment_utt = [ 1, 1, 3, 3, 4, 5, 7, 7, 9, 10, 11, 12, 13, 15, 17, 17, 19, 21, 23, 23]
+    
+    in which case the output would be:
+    boundary_info_utt = [
+        {'text': 'hi', 's_start': 1, 's_end': 3, 't_start': 0, 't_end': 3},
+        {'text': 'world', 's_start': 7, 's_end': 15, 't_start': 6, 't_end': 13},
+        {'text': 'hey', 's_start': 19, 's_end': 23, 't_start': 16, 't_end': 19},
+    ]
+    """
+    # first remove boundary_info of any items that are not in the alignment
+    # the only items we expect not to be in the alignment are blanks that the alignment chooses to skip
+    # we will iterate boundary_info in reverse order for this to make popping the items simple
+    s_in_alignment = set(alignment_utt)
+    for boundary_info_pointer in range(len(boundary_info_utt) - 1, -1, -1):
+        s_in_boundary_info = set(
+            range(
+                boundary_info_utt[boundary_info_pointer]["s_start"],
+                boundary_info_utt[boundary_info_pointer]["s_end"] + 1,
+            )
+        )
+        item_not_in_alignment = True
+        for s_ in s_in_boundary_info:
+            if s_ in s_in_alignment:
+                item_not_in_alignment = False
+
+        if item_not_in_alignment:
+            boundary_info_utt.pop(boundary_info_pointer)
+
+    # now update boundary_info with t_start and t_end
+    boundary_info_pointer = 0
+    for t, s_at_t in enumerate(alignment_utt):
+        if s_at_t == boundary_info_utt[boundary_info_pointer]["s_start"]:
+            if "t_start" not in boundary_info_utt[boundary_info_pointer]:
+                # we have just reached the start of the word/token/segment in the alignment => update t_start
+                boundary_info_utt[boundary_info_pointer]["t_start"] = t
+
+        if t < len(alignment_utt) - 1:  # this if is to avoid accessing an index that is not in the list
+            if alignment_utt[t + 1] > boundary_info_utt[boundary_info_pointer]["s_end"]:
+                if "t_end" not in boundary_info_utt[boundary_info_pointer]:
+                    boundary_info_utt[boundary_info_pointer]["t_end"] = t
+
+                boundary_info_pointer += 1
+        else:  # i.e. t == len(alignment) - 1, i.e. we are a the final element in alignment
+            # add final t_end if we haven't already
+            if "t_end" not in boundary_info_utt[boundary_info_pointer]:
+                boundary_info_utt[boundary_info_pointer]["t_end"] = t
+
+        if boundary_info_pointer == len(boundary_info_utt):
+            # we have finished populating boundary_info with t_start and t_end,
+            # but we might have some final remaining elements (blanks) in the alignment which we dont care about
+            # => break, so as not to cause issues trying to access boundary_info[boundary_info_pointer]
+            break
+
+    return boundary_info_utt
+
+
+def make_ctm(
```

Comment:
Add data types to signature in the future or pass around a dataclass as a datastructure. This is way too complex.